File: backend-elixir/lib/backend_elixir/rest/error_mapper.ex
defmodule BackendElixir.Rest.ErrorMapper do
  @moduledoc """
  backend-java/backend-kotlin/backend-python/backend-rustのRestErrorMapper/status_and_bodyと
  1文字も変えていないJSON形状・HTTPステータス(CONTRACT.mdセクション20.5のワイヤー契約パリティ)
  """

  alias BackendElixir.Domain.TaskError

  def status_and_body(%TaskError{kind: :validation, message: message}) do
    {422, %{error: "validation_error", message: message}}
  end

  def status_and_body(%TaskError{kind: kind}) do
    {status, error_key} = kind_to_status_and_error(kind)
    {status, %{error: error_key}}
  end

  defp kind_to_status_and_error(:unauthorized), do: {401, "unauthorized"}
  defp kind_to_status_and_error(:unauthenticated), do: {401, "unauthenticated"}
  defp kind_to_status_and_error(:user_not_provisioned), do: {403, "user_not_provisioned"}
  defp kind_to_status_and_error(:invalid_request), do: {400, "invalid_request"}
  defp kind_to_status_and_error(:invalid_id), do: {400, "invalid_id"}
  defp kind_to_status_and_error(:user_id_required), do: {400, "user_id_required"}
  defp kind_to_status_and_error(:invalid_user_id), do: {400, "invalid_user_id"}
  defp kind_to_status_and_error(:invalid_status), do: {422, "invalid_status"}
  defp kind_to_status_and_error(:invalid_finished_on), do: {422, "invalid_finished_on"}
  defp kind_to_status_and_error(:not_found), do: {404, "not_found"}
  defp kind_to_status_and_error(:internal), do: {500, "internal_server_error"}
end

File: backend-haskell/src/BackendHaskell/Rest/ErrorMapper.hs
-- | backend-java/backend-kotlin/backend-python/backend-elixirのErrorMapper/status_and_bodyと
-- 1文字も変えていないJSON形状・HTTPステータス(CONTRACT.mdセクション20.5のワイヤー契約パリティ)
module BackendHaskell.Rest.ErrorMapper
  ( statusAndBody
  ) where

import Data.Aeson (Value, object, (.=))
import Data.Text (Text)

import BackendHaskell.Domain.Errors (TaskError (..), TaskErrorKind (..))

statusAndBody :: TaskError -> (Int, Value)
statusAndBody (TaskError Validation message) = (422, object ["error" .= ("validation_error" :: Text), "message" .= message])
statusAndBody (TaskError kind _) = kindToStatusAndError kind

kindToStatusAndError :: TaskErrorKind -> (Int, Value)
kindToStatusAndError Unauthorized = (401, object ["error" .= ("unauthorized" :: Text)])
kindToStatusAndError UserNotProvisioned = (403, object ["error" .= ("user_not_provisioned" :: Text)])
kindToStatusAndError InvalidRequest = (400, object ["error" .= ("invalid_request" :: Text)])
kindToStatusAndError InvalidId = (400, object ["error" .= ("invalid_id" :: Text)])
kindToStatusAndError InvalidStatus = (422, object ["error" .= ("invalid_status" :: Text)])
kindToStatusAndError InvalidFinishedOn = (422, object ["error" .= ("invalid_finished_on" :: Text)])
kindToStatusAndError NotFound = (404, object ["error" .= ("not_found" :: Text)])
kindToStatusAndError Internal = (500, object ["error" .= ("internal_server_error" :: Text)])
kindToStatusAndError Validation = (422, object ["error" .= ("validation_error" :: Text)])
